pkg/mover/chart.go

```diff
@@ -253,3 +276,19 @@ func saveChart(chart *chart.Chart, toChartFilename string) error {
 
 	return os.Remove(tempDir)
 }
+
+type Option func(*ChartMover)
+
+// WithRetries defines how many times to retry the push operation
+func WithRetries(retries uint) Option {
+	return func(c *ChartMover) {
+		c.retries = retries
+	}
+}
```

Comment:
An alternative could be:

~~~go
// WithRetries defines how many times to retry the push operation
func (c *chartMover) WithRetries(retries uint) (c *chartMover)  {
  c.retries = retries
  return c
}
~~~

And it will be invoked such as:
```go
chartMover, err := mover.NewChartMover(
		sourceChart,
		imagePatterns,
		rules
).WithRetries(Retries).WithLogger(cmd)
```

I believe that is more idiomatic. It resembles the Context interface for instance.

Reply:
I guess that the issue with your approach is that you are constructing the object and mutating it, while my approach builds it using the provided options. And in our case, the constructor does more than it should, i.e pulling images and to do so we might need a logger and a retry number. Which means that we will need all the context during the initial construction.

About using them as parameters of a constructor, as I mentioned, I've seen it before i.e https://pkg.go.dev/google.golang.org/grpc#NewServer or https://github.com/helm/helm/blob/179f90151d5ecb4aa3d35ada35e82b5c1e791752/pkg/getter/httpgetter_test.go#L55

re: Context interface. I am not sure that pattern is a good fit, since WIthTimeout (for example) is actually a function, not a method, and it takes as argument a parent context, creating a completely new instance.

So I am not sure if it's a good idea to move to the model you described, let me know what you think or if I am missing smth

Reply:
You are right! Yes in this case my alternative would not work as the "constructor" is doing more work than just constructing.